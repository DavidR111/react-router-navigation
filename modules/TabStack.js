/* @flow */
/* eslint no-duplicate-imports: 0 */

import { Component } from 'react'
import { withRouter } from 'react-router'
import { StateUtils } from 'react-navigation'
import type { RouterHistory, Location } from 'react-router'
import type { NavigationState, TabRendererProps, Tabs, TabRoute, TabProps } from './TypeDefinitions'
import StackUtils from './StackUtils'

type Props = RouterHistory & {
  children: Array<React$Element<TabProps>>,
  render: (props: TabRendererProps) => React$Element<any>,
  forceSync?: boolean,
}

type State = {
  navigationState: NavigationState<TabRoute>,
  tabs: Tabs,
  rootIndex: number,
  history: { [key: number]: Array<Location> },
}

class TabStack extends Component<void, Props, State> {

  props: Props
  state: State

  // Initialyze navigation state with initial history
  constructor(props: Props): void {
    super(props)
    // Build the tab stack ($FlowFixMe)
    const { children, entries, location } = props
    const tabs = StackUtils.build(children)
    // Get initial route
    const currentRoute = StackUtils.getRoute(tabs, location)
    if (!currentRoute) throw new Error('No route found !')
    // Build navigation state
    const routes = tabs.map((tab) => ({
      // $FlowFixMe
      key: StackUtils.createKey({ key: tab.key }),
      routeName: tab.path,
    }))
    const index = routes.findIndex(({ routeName }) => currentRoute.key === routeName)
    const navigationState = { index, routes }
    const rootIndex = props.index || 0
    // Initialyze cached history
    const history = {
      [index]: entries.slice(location.index),
    }
    // Save everything
    this.state = { navigationState, tabs, rootIndex, history }
  }

  // Listen all history events
  componentWillReceiveProps(nextProps): void {
    // Get current route ($FlowFixMe)
    const { location, entries } = nextProps
    const { navigationState: { routes, index }, tabs, rootIndex } = this.state
    // Get current tab
    const currentRoute = routes[index]
    const currentTab = StackUtils.get(tabs, currentRoute)
    // Get next tab
    const nextRoute = StackUtils.getRoute(tabs, location)
    if (!nextRoute) return
    const nextTab = StackUtils.get(tabs, nextRoute)
    const nextIndex = routes.findIndex(({ routeName }) => routeName === nextRoute.key)
    // Update navigation state
    if (
      currentTab && nextTab &&
      StackUtils.shouldUpdate(currentTab, nextTab, this.props, nextProps)
    ) {
      this.setState((state) => ({
        navigationState: StateUtils.jumpToIndex(
          state.navigationState,
          nextIndex,
        ),
      }))
    }
    // Update history
    if (nextRoute) {
      this.state.history[nextIndex] = entries.slice(
        rootIndex,
        nextProps.index + 1,
      )
    }
  }

  // Callback for when the current tab changes
  onRequestChangeTab = (index: number): void => {
    const entries = this.state.history[index]
    const tab = this.state.tabs[index]
    if (tab) {
      if (this.props.forceSync) {
        const n = this.state.rootIndex - (this.props.index || 0)
        this.props.go(n)
        this.props.replace(tab.path)
        if (entries) {
          entries
            .slice(this.state.rootIndex + 1)
            .forEach(({ pathname }) => {
              this.props.push(pathname)
            })
          this.props.replace(
            entries[Math.max(0, parseInt(entries.length - 1))].pathname
          )
        }
      } else {
        this.props.replace(tab.path)
      }
    }
  }

  // Render view
  render(): React$Element<any> {
    return this.props.render({
      ...this.state,
      onRequestChangeTab: this.onRequestChangeTab,
    })
  }

}

export default withRouter(TabStack)
