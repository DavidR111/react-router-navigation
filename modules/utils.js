/* @flow */
/* eslint no-param-reassign: 0 */

import { Children } from 'react'
import { matchPattern } from 'react-router'
import type { NavigationRoute } from 'react-native/Libraries/NavigationExperimental/NavigationTypeDefinition'
import type { Card, MatchCardProps, Tab, MatchTabProps } from './StackTypeDefinitions'

/**
 * Extract current match from all matches
 * and return a route with key prop
 */
export function getCurrentRoute(
  matchs: Array<{ pattern: string, exactly?: boolean } & Object>,
  parent: ?string,
  location: { pathname: string },
): ?NavigationRoute {
  const route = matchs
    .find(({ pattern, exactly }) => {
      return matchPattern(pattern, location, exactly || false, parent)
    })
  if (!route) return null
  return { key: route.pattern }
}


/**
 * Build an item stack by extracting
 * all valid props
 */
export function buildItemStack<Item>(
  children: Array<React$Element<Item>>,
  advancedProps: Array<string>,
  matchComponent: string,
): Array<Item & { key: string }> {
  // Define required props
  const defaultProps = ['component', 'render', 'pattern', 'exactly']
  // Map all children
  return Children.toArray(children).reduce((items, child) => {
    // child === <Match /> || child === <MatchCard />
    const displayName = child.displayName || child.type.displayName || child.type.name
    const isAdvancedMatch = matchComponent === displayName
    if (displayName !== 'Match' && !isAdvancedMatch) return items
    // Build item
    const item = Object.keys(child.props).reduce((props, key) => {
      if (key === 'pattern') {
        props.key = child.props[key]
        props[key] = child.props[key]
      } else if (isAdvancedMatch && (key === 'render' || key === 'component')) {
        props[key] = () => child
      } else if (
        defaultProps.includes(key) ||
        (isAdvancedMatch && advancedProps.includes(key))
      ) {
        props[key] = child.props[key]
      }
      return props
    }, {})
    if (!item.key) return items
    return [...items, item]
  }, [])
}


/**
 * Build a card stack
 */
export function buildCards(children: Array<React$Element<MatchCardProps>>): Array<Card> {
  return buildItemStack(children, ['title', 'hideNavBar'], 'MatchCard')
}


/**
 * Build a tab stack
 */
export function buildTabs(children: Array<React$Element<MatchTabProps>>): Array<Tab> {
  return buildItemStack(children, ['label', 'tabBarStyle'], 'MatchTab')
}
