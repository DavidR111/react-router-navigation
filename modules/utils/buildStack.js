/* @flow */
/* eslint no-param-reassign: 0*/

import { Children, cloneElement } from 'react'

export default function buildStack<Item>(
  children: Array<React$Element<Item>>
): Array<Item & { key: string }> {
  return Children.toArray(children).reduce((stack, child) => {
    const item = Object.keys(child.props).reduce((props, key) => {
      if (key === 'path') {
        props.key = child.props[key]
        props[key] = child.props[key]
      } else if (key === 'render' || key === 'component' || key === 'children') {
        const renderMethod = (ownProps: any) => cloneElement(child, ownProps)
        Object.defineProperty(renderMethod, 'name', { value: key })
        props[key] = renderMethod
      } else {
        props[key] = child.props[key]
      }
      return props
    }, {})
    if (!item.path) return stack
    return [...stack, item]
  }, [])
}
