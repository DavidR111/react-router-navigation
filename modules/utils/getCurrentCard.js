/* @flow */

import type { Route, Cards, Card } from './../TypeDefinitions'
import normalizeRoute from './normalizeRoute'

export default function getCurrentCard(cards: Cards, route: Route): ?Card {
  return cards.find((card) => {
    return normalizeRoute(route).key === card.key
  })
}
