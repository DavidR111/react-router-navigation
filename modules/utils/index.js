/* @flow */

export { default as buildStack } from './buildStack'
export { default as getCurrentCard } from './getCurrentCard'
export { default as getCurrentRoute } from './getCurrentRoute'
export { default as normalizeRoute } from './normalizeRoute'
export { default as shouldStackUpdate } from './shouldStackUpdate'
