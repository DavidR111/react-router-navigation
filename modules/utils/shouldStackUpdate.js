/* @flow */
/* eslint no-duplicate-imports: 0*/

import { matchPath } from 'react-router'
import type { RouterHistory, Location } from 'react-router'
import type { Card } from './../TypeDefinitions'

export default function (
  currentCard: Card,
  nextCard: Card,
  routerHistory: RouterHistory,
  previousLocation: Location,
): boolean {
  const { entries, location, index } = routerHistory
  if (entries === undefined || index === undefined) return false
  const previousEntry = entries[index - 1]
  const currentEntry = entries[index]
  const nextEntry = entries[index + 1]
  const matchCurrentRoute = matchPath(location.pathname, currentCard.path, currentCard)
  const matchNextRoute = matchPath(location.pathname, nextCard.path, nextCard)
  return (
    // URL changes
    (previousLocation.pathname !== location.pathname) &&
    // case 1) basic pathname
    ((currentCard.key !== nextCard.key) ||
    // case 2) Pathname with query params
    // Ex: with same path article/:id,
    //     pathname article/2 !== article/3
    (matchCurrentRoute !== null && matchNextRoute !== null &&
     Object.keys(matchCurrentRoute.params).length !== 0 &&
     Object.keys(matchNextRoute.params).length !== 0 &&
     ((previousEntry && currentEntry.pathname !== previousEntry.pathname) ||
      (nextEntry && currentEntry.pathname !== nextEntry.pathname))
    ))
  )
}
