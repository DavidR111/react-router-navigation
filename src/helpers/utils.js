/* @flow */

import _ from 'lodash'
import type { NavigationState, NavigationRoute, NavigationScene } from './../types'

export function extractScenes(
  children: Array<React$Element<NavigationScene>>
): Array<NavigationRoute> {
  return children.map((child) => ({
    ...child.props,
    key: child.key,
  }))
}

export function normalizePath(path: string): string {
  return path
    .split('.')
    .filter((_path) => _path)
    .map((_path) => `routes[${_path}]`)
    .join('.')
}

export function getSiblingScenes(state: NavigationState) {
  const normalizedPath = normalizePath(state.path.slice(0, -2))
  const path = normalizedPath
    ? `${normalizedPath}.children`
    : 'children'
  const sceneChildren = _.get(state, path)
  return extractScenes(sceneChildren)
}

export function getCurrentRoute(state: NavigationState): NavigationRoute {
  const normalizedPath = normalizePath(state.path)
  return _.get(state, normalizedPath)
}

export function findPathOfClosestTabs(state: NavigationState): string {
  const paths = state.path.split('.')
  const path = paths
    .map((_path, index) => paths.slice(0, index + 1).join('.'))
    .reverse()
    .map((_path) => ({
      path: _path,
      navigationState: _.get(state, normalizePath(_path)),
    }))
    .filter(({ navigationState }) => navigationState.tabs)
    .map((tabState) => tabState.path)
    .join('')
  return path
}
