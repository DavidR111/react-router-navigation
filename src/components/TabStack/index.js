/* @flow */

import React, { PropTypes, Component } from 'react'
import _ from 'lodash'
import CardStack from './../CardStack'
import TabView from './../TabView'
import { normalizePath } from './../../utils'
import type { NavigationState, NavigationSceneProps } from './../../types'

type Props = {
  navigationState: NavigationState,
  renderScene: (sceneProps: NavigationSceneProps) => React$Element<any>,
  renderHeader: (sceneProps: NavigationSceneProps) => React$Element<any> | null,
  push: (key: string, callback: Function) => void,
  pop: (callback: Function) => void,
  changeTab: (index: number, callback: Function) => void,
}

type DefaultProps = {
  renderHeader: () => null,
}

type State = {
  navigationState: NavigationState,
}

class TabStack extends Component {

  state: State
  props: Props

  static defaultProps: DefaultProps = {
    renderHeader: () => null,
  }


  // Clone navigation state to it's own
  // in this component and remove circular
  // references
  constructor(props: Props) {
    super(props)
    const { navigationState } = props
    const path = normalizePath(navigationState.path.slice(0, -4))
    const extractedNavigationState = path
      ? _.cloneDeep(_.get(navigationState, path))
      : navigationState
    this.state = {
      navigationState: extractedNavigationState,
    }
  }


  // Dispatch actions
  push = (key: string) => {
    this.props.push(key, this.updateNavigationState)
  }
  changeTab = (index: number) => {
    this.props.changeTab(index, this.updateNavigationState)
  }
  pop = () => {
    this.props.pop(this.updateNavigationState)
  }


  // Update local navigation state each
  // time an action is dispatched
  updateNavigationState = (navigationState: NavigationState): void => {
    const path = normalizePath(navigationState.path.slice(0, -4))
    const nextNavigationState = path
      ? _.cloneDeep(_.get(navigationState, path))
      : navigationState
    this.setState({ navigationState: nextNavigationState })
  }


  // Provided updated router context
  static childContextTypes = {
    router: PropTypes.object.isRequired,
  }
  getChildContext() {
    return {
      router: {
        push: this.push,
        pop: this.pop,
      },
    }
  }


  // Render each tab in a <CardStack />
  // component
  render() {
    return (
      <TabView
        navigationState={this.state.navigationState}
        changeTab={this.changeTab}
        renderScene={(navigationState) => (
          <CardStack
            navigationState={navigationState}
            pop={this.pop}
            renderScene={this.props.renderScene}
            renderHeader={(sceneProps) => this.props.renderHeader(sceneProps, this.pop)}
          />
        )}
      />
    )
  }

}

export default TabStack
