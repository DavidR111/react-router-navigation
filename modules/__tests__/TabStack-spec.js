/* eslint new-cap: 0 */

import React from 'react'
import { TouchableOpacity } from 'react-native'
import { Router, Route } from 'react-router'
import createHistory from 'history/createMemoryHistory'
import renderer from 'react-test-renderer'
import './__mocks__'
import { componentFactory, TabView } from './utils'
import TabStack from './../TabStack'

it('<TabStack /> renders correctly', () => {
  const history = createHistory()
  const component = renderer.create(
    <Router history={history}>
      <Route>
        {routeProps => (
          <TabStack {...routeProps} render={TabView}>
            <Route exact path="/" render={componentFactory('Index')} />
            <Route path="/hello" render={componentFactory('Hello')} />
            <Route path="/goodbye" render={componentFactory('Goodbye')} />
          </TabStack>
        )}
      </Route>
    </Router>,
  )
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

it('<TabStack /> renders correctly with initialEntries prop ', () => {
  const history = createHistory({ initialEntries: ['/hello'] })
  const component = renderer.create(
    <Router history={history}>
      <Route>
        {routeProps => (
          <TabStack {...routeProps} render={TabView}>
            <Route exact path="/" render={componentFactory('Index')} />
            <Route path="/hello" render={componentFactory('Hello')} />
            <Route path="/goodbye" render={componentFactory('Goodbye')} />
          </TabStack>
        )}
      </Route>
    </Router>,
  )
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

it('<TabStack /> re-renders correctly when "replace" action is called', () => {
  const history = createHistory()
  const component = renderer.create(
    <Router history={history}>
      <Route>
        {routeProps => (
          <TabStack {...routeProps} render={TabView}>
            <Route exact path="/" render={componentFactory('Index')} />
            <Route path="/hello" render={componentFactory('Hello')} />
            <Route path="/goodbye" render={componentFactory('Goodbye')} />
          </TabStack>
        )}
      </Route>
    </Router>,
  )
  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()
  history.replace('/hello')
  tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

it('<TabStack /> re-renders correctly when onRequestChangeTab() method is called', () => {
  const history = createHistory()
  const component = renderer.create(
    <Router history={history}>
      <Route>
        {routeProps => (
          <TabStack
            {...routeProps}
            render={({ navigationState, onRequestChangeTab, tabs }) => (
              <TouchableOpacity onPress={onRequestChangeTab}>
                {TabView({ navigationState, onRequestChangeTab, tabs })}
              </TouchableOpacity>
            )}
          >
            <Route exact path="/" render={componentFactory('Index')} />
            <Route path="/hello" render={componentFactory('Hello')} />
            <Route path="/goodbye" render={componentFactory('Goodbye')} />
          </TabStack>
        )}
      </Route>
    </Router>,
  )
  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()
  tree.props.onPress(2)
  tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})
