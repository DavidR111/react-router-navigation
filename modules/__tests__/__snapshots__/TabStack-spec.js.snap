exports[`test <TabStack /> re-renders correctly when "replace" action is called 1`] = `
<Text>
  Index
</Text>
`;

exports[`test <TabStack /> re-renders correctly when "replace" action is called 2`] = `
<Text>
  Hello
</Text>
`;

exports[`test <TabStack /> renders correctly 1`] = `
<Text>
  Index
</Text>
`;

exports[`test <TabStack /> renders correctly with initialEntries prop  1`] = `
<Text>
  Hello
</Text>
`;
